src/java/org/apache/cassandra/cql3/selection/SortedRowsBuilder.java

```diff
@@ -209,4 +219,210 @@ private RowWithScore(List<ByteBuffer> row, float score)
             }
         }
     }
+
+    /**
+     * {@link SortedRowsBuilder} that orders rows based on the provided comparator.
+     * </p>
+     * It's possible for the comparison to produce ties. To deal with these ties, the rows are decorated with their
+     * position in the sequence of calls to {@link #add(List)}, so we can use that identifying position to solve ties by
+     * favoring the row that was inserted first.
+     * </p>
+     * The rows can be decorated with any other value used for the comparator, so it doesn't need to recalculate that
+     * value in every comparison.
+     * </p>
+     * It keeps at most {@code limit + offset} rows in memory.
+     */
+    public static class WithHeapSort<T extends WithHeapSort.RowWithId> extends SortedRowsBuilder
+    {
+        private final BiFunction<List<ByteBuffer>, Integer, T> decorator;
+        private final PriorityQueue<T> heap;
+        private final int heapCapacity;
+
+        private List<T> initialRows = new ArrayList<>(); // first limit+offset rows to be added with PriorityQueue#addAll
```

Comment:
is this structure accessed only by 1 thread ?

Reply:
Yes, it is. The code that this replaces used an `ArrayList` to do the sorting.